File: 0076-minimum-window-substring/0076-minimum-window-substring.py
class Solution:
    def minWindow(self, s: str, t: str) -> str:
        maxLen=10**9
        Hash=[0]*256

        for c in t:
            Hash[ord(c)]+=1
        
        n,m=len(s),len(t)
        sidx=n
        l=0
        r=0
        count=0
        while(r<n):
            if Hash[ord(s[r])]>0:
                count+=1
            Hash[ord(s[r])]-=1
            while(count==m):
                if r-l+1<maxLen:
                    maxLen=r-l+1
                    sidx=l
                Hash[ord(s[l])]+=1
                if(Hash[ord(s[l])])>0:
                    count-=1
                l+=1
            r+=1
        return s[sidx:sidx+maxLen] 

File: 0076-minimum-window-substring/0076-minimum-window-substring.java
import java.util.*;
class Solution {
    public String minWindow(String s, String t) {
        int left=0;
        int right=0;
        int count=0;
        int minLen=Integer.MAX_VALUE;
        int m=t.length();
        int n=s.length();
        int startIndex=-1;
        Map<Character,Integer> map=t.chars()
                                    .mapToObj(c->(char)c)
                                    .collect(Collectors.groupingBy(
                                        Function.identity(),
                                        Collectors.summingInt(c->1)
                                    ));
        while(right<n){
            if (map.getOrDefault(s.charAt(right),0)>0){
                count++;
            }
            map.put(s.charAt(right),map.getOrDefault(s.charAt(right),0)-1);
            while(count==m){

                if(right-left+1<minLen){
                    minLen=right-left+1;
                    startIndex=left;
                }
                map.put(s.charAt(left),map.getOrDefault(s.charAt(left),0)+1);
                if(map.getOrDefault(s.charAt(left),0)>0){
                    count--;
                }
                left++;
            }
            right++;
        }
        return (startIndex == -1) ? "" :s.substring(startIndex,startIndex+minLen);
    }
}